internal/provider/resource_credential.go

```diff
@@ -217,44 +315,118 @@ func resourceCredentialUpdate(ctx context.Context, d *schema.ResourceData, meta
+
+	// Read Terraform configuration from the request into the resource model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Access prior state data
 
-	id := strings.Split(d.Id(), ":")
-	if len(id) != 4 {
-		return diag.Errorf("unable to parse credential name and cloud name from provided ID")
+	resID := retrieveValidateID(data, &resp.Diagnostics, "update")
+	if resp.Diagnostics.HasError() {
+		return
 	}
-	credentialName, cloudName, clientCredentialStr, controllerCredentialStr := id[0], id[1], id[2], id[3]
+	// extract : name & cloud.name, client_credential, controller_credential
+	credentialName, cloudName, clientCredentialStr, controllerCredentialStr := resID[0], resID[1], resID[2], resID[3]
 	clientCredential, controllerCredential, err := convertOptionsBool(clientCredentialStr, controllerCredentialStr)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Provider Error", err.Error())
+		return
 	}
 
-	err = client.Credentials.DestroyCredential(juju.DestroyCredentialInput{
+	// Perform external call to destroy the resource
+	err = c.client.Credentials.DestroyCredential(juju.DestroyCredentialInput{
 		ClientCredential:     clientCredential,
 		CloudName:            cloudName,
 		ControllerCredential: controllerCredential,
 		Name:                 credentialName,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete credential resource, got error: %s", err))
+	}
+	tflog.Trace(ctx, fmt.Sprintf("deleted credential resource %q", credentialName))
+}
+
+func (c *credentialResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
 	}
+	c.client = client
+}
+
+func (c credentialResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
 
-	d.SetId("")
+func convertRawAttributes(attributesRaw map[string]attr.Value) map[string]string {
+	newAttributes := make(map[string]string)
+	for key, value := range attributesRaw {
+		newAttributes[key] = attributeEntryToString(valueAttrToString(value))
+	}
+	return newAttributes
+}
+
+func newIDFrom(credentialName string, cloudName string, clientCredential bool, controllerCredential bool) string {
+	return fmt.Sprintf("%s:%s:%t:%t", credentialName, cloudName, clientCredential, controllerCredential)
+}
+
+func retrieveValidateID(model *credentialResourceModel, diag *diag.Diagnostics, method string) []string {
```

Comment:
update the function name to include reference to credentials, it is specific to credential IDs.